File: src/main/kotlin/ru/digital/image/picture/images/MinMaxColorOfPointGetter.java
package ru.digital.image.picture.images;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class MinMaxColorOfPointGetter implements OnePointColorGetter {


    public MinMaxColorOfPointGetter(int[] defaultColors){
        this.defaultColors = Arrays.stream(defaultColors).mapToObj(
                it -> new RGB(it)
        ).collect(Collectors.toList());
    }
    @Override
    public int get(int color) {
        var current = new RGB(color);
        Object[] obj = defaultColors.stream().map(
                it ->
                        new Object[]{it, it.max(current)}
        ).min(
                (it1, it2) -> {
                    if ((double) it1[1] > (double) it2[1])
                        return 1;
                    else if ((double) it1[1] == (double) it2[1])
                        return 0;
                    else return -1;
                }
        ).get();
        return ((RGB) obj[0]).toInt();
    }

    private List<RGB> defaultColors;
}

File: src/main/kotlin/ru/digital/image/picture/images/changer/MinMaxColorOfPointGetter.kt
package ru.digital.image.picture.images.changer

import ru.digital.image.picture.images.domain.RGB
import java.util.*
import java.util.function.Function
import java.util.stream.Collectors

class MinMaxColorOfPointGetter(defaultColors: IntArray?) : PointColorGetter {
    operator fun get(color: Int): Int {
        val current = RGB(color)
        val obj = defaultColors.stream().map { it: RGB ->
            arrayOf(
                it,
                it.max(current)
            )
        }.min { it1: Array<Any>, it2: Array<Any> ->
            if (it1[1] as Double > it2[1] as Double
            ) return@min 1 else if (it1[1] as Double == it2[1] as Double) return@min 0 else return@min -1
        }.get()
        return (obj[0] as RGB).toInt()
    }

    private val defaultColors: List<RGB>

    init {
        this.defaultColors =
            Arrays.stream(defaultColors).mapToObj<Any> { it: Int -> RGB(it) }.collect(Collectors.toList()) as List<RGB>
    }
}
